src/core/io/src/4C_io_input_spec_builders.hpp

```diff
@@ -0,0 +1,884 @@
+     * // size is given as a callback.
+     * entry<int>("N");
+     * entry<std::vector<double>>("my_vector", {.size = from_parameter<int>("N")});
+     * @endcode
+     *
+     * After parsing an InputSpec with fully_parse(), the value of the entry can be retrieved from
+     * an InputParameterContainer. The details depend on the `required` and `default_value` fields:
+     *
+     *   - `required` is used to determine whether the parameter is required in the input file. Note
+     *     that by default, if `required` is not set, the parameter is implicitly required. Failing
+     *     to read a required parameter will result in an exception. On the other hand, if a
+     *     parameter is not required, it is optional and can be left out in the input file. The
+     *     container that is filled by the fully_parse() function will not contain the parameter.
+     *
+     *   - `default_value` is used to provide a value that is used if the parameter is not present
+     *     in the input file. If `default_value` is set, the parameter is implicitly optional. If
+     *     the parameter is not present in the input file, the default value will be stored in the
+     *     container that is filled by the fully_parse() function.
+     *
+     *   - Setting both `required = true` and a `default_value` is a logical error and will result
+     *     in an exception.
+     *
+     * When you decide how to set the `required` and `default_value` fields, consider the following
+     * three cases:
+     *
+     *   - If you always require a parameter and there is no reasonable default value, do not set
+     *     `required` or `default_value`. This will make the parameter required by default. Parsing
+     *     will fail if the parameter is not present, but after parsing you can be sure that the
+     *     parameter can safely be retrieved with InputParameterContainer::get(). A good example is
+     *     the time step size in a time integration scheme: this parameter is always required and
+     *     taking an arbitrary default value is not a good idea.
+     *
+     *   - Is there a reasonable default value for the parameter, which works in most situations? If
+     *     yes, set `default_value` to this value (`required` implicitly is `false` then). This
+     *     guarantees that you can always read the a value from the container with
+     *     InputParameterContainer::get(). A good example is a parameter that activates or
+     *     deactivates a feature, e.g., defaulting the EAS element technology to off might be
+     *     reasonable.
+     *
+     *   - If the parameter is not required, but there is no reasonable default value, set
+     *     `required = false`. This makes the parameter optional and the container might not contain
+     *     the parameter if it is not present in the input file. Use
+     *     InputParameterContainer::get_if() or InputParameterContainer::get_or() to safely retrieve
```

Comment:
Do we really want to provide the `get_or()`? To me, it looks like it's a silent default.

Reply:
Yes, if you do this in your code, it is a silent default. It depends on whether you think about the `InputParameterContainer` in isolation, where `get_or` makes sense, or if you think about it in combination with `InputSpec`, which would provide a different way to specify defaults. `get_or` predates `InputSpec`. I am not 100% sure  what we will need, but I suspect you may be right and `get_or` should not be necessary in combination with `InputSpec`.